Add timeout and tests for curb, also added the gem curb that was not part of selenium

aguspe opened the pull request:
### **User description**
### Description
Based on #14121 we wanted to regain the ability to set timeout in our curb clients, to do this I added the attribute accessor timeout

Besides this tests were added and the gem curb was added as a dependency because it was not present in Selenium

### Motivation and Context
It's important to provide users with different clients that they can set up and configure with the correct timeouts

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.


___

### **PR Type**
Enhancement, Tests


___

### **Description**
- Added `timeout` attribute accessor to `Curb` HTTP client.
- Modified `initialize` method in `Curb` to accept `timeout`.
- Updated `client` method in `Curb` to set `timeout` if provided.
- Added `curb` gem as a dependency in the gemspec.
- Added unit tests for `Curb` HTTP client, including tests for default and custom timeout values.
- Added type signatures for `Curl::Easy` class and `Curb` class methods and attributes.



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>remote.rb</strong><dd><code>Add autoload for Curb in Http module</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/lib/selenium/webdriver/remote.rb

- Added autoload for `Curb` in the `Http` module.



</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14285/files#diff-8508e5626045769b4fff77bdc225df7fc68b5d086c29c7566ec62232604ad5a7">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>curb.rb</strong><dd><code>Add timeout handling to Curb HTTP client</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/lib/selenium/webdriver/remote/http/curb.rb

<li>Added <code>timeout</code> attribute accessor.<br> <li> Modified <code>initialize</code> method to accept <code>timeout</code>.<br> <li> Updated <code>client</code> method to set <code>timeout</code> if provided.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14285/files#diff-eb552196032baf211280003920bc4d7eed1d0571efe05b13522276ddc474ddd5">+14/-6</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>curl.rbs</strong><dd><code>Add type signature for Curl::Easy class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/sig/gems/curb/curl.rbs

- Added type signature for `Curl::Easy` class.



</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14285/files#diff-bf99b7027a15b3cbfc08d59f8e7be1a7607cc2256a4c83e7026ca48246020cca">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>curb.rbs</strong><dd><code>Add type signatures for Curb class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/sig/lib/selenium/webdriver/remote/http/curb.rbs

- Added type signatures for `Curb` class methods and attributes.



</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14285/files#diff-c1f7c202438feed70107ad30f6813d7fbdfc2957dbfa859780e666325b2fc7a2">+7/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Dependencies</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>selenium-webdriver.gemspec</strong><dd><code>Add curb gem dependency</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/selenium-webdriver.gemspec

- Added `curb` gem as a dependency.



</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14285/files#diff-f7c3979ae889cfb6e1b2386569a8bcd4f45e9a7e03d8c13bad52eb502a240aad">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>curb_spec.rb</strong><dd><code>Add unit tests for Curb HTTP client</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/spec/unit/selenium/webdriver/remote/http/curb_spec.rb

<li>Added unit tests for <code>Curb</code> HTTP client.<br> <li> Tested default and custom timeout values.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14285/files#diff-33589188c6e59a0827eee131e4ccbac06900d395edc224b524d9149ea3b8d717">+51/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

p0deje commented on rb/selenium-webdriver.gemspec:
```diff
@@ -48,6 +48,7 @@ Gem::Specification.new do |s|
   s.require_paths = ['lib']
 
   s.add_dependency 'base64', ['~> 0.2']
+  s.add_dependency 'curb', ['~> 1.0.5']
```
That was by design, the idea is that people can add `curb` gem to their own Gemfile if they want to use this HTTP client.

aguspe replied:
Oh great thank you for the explanation :)

p0deje replied:
For the purpose of this PR, it would make sense to add `curb` as a development dependency so you could run specs for it.

Also, once Gemfile.lock is updated, please run `bazel build @bundle` to regenerate the checksums. The command should print `gem_checksums = {...}` to the output which you can then copy/paste to MODULE.bazel.

p0deje commented on rb/lib/selenium/webdriver/remote.rb:
```diff
@@ -31,7 +31,8 @@ module Remote
       module Http
-        autoload :Common,  'selenium/webdriver/remote/http/common'
+        autoload :Common, 'selenium/webdriver/remote/http/common'
+        autoload :Curb, 'selenium/webdriver/remote/http/curb'
```
Can you also update the `@example` in the client and remove unnecessary `require/include`?

diemol commented:
Seems something is failing in the build process?
https://github.com/SeleniumHQ/selenium/actions/runs/10305049245/job/28524950266?pr=14285#step:15:2484

aguspe commented:
> Details

I will look into the failed-to-build tests tomorrow and also I will add the checksum update that I'm missing, thank you for executing the pipeline!

aguspe commented:
@p0deje I ran ```bazel build @bundle```but I cannot find where the checksums of the gems are printed, do you know where should I be looking for them?

p0deje commented:
@aguspe Can you run the following and share the output? Note that it's going to take a while.

```
bazel clean --expunge # clear all Bazel caches
bazel build @bundle
```

aguspe commented:
> bazel clean --expunge # clear all Bazel caches

Thank you so much @p0deje, that made the trick, the check sums are updated, so it's ready for re-review

p0deje commented on rb/selenium-webdriver.gemspec:
```diff
@@ -53,6 +53,7 @@ Gem::Specification.new do |s|
   s.add_dependency 'rubyzip', ['>= 1.2.2', '< 3.0']
   s.add_dependency 'websocket', ['~> 1.0']
 
+  s.add_dependency 'curb', ['~> 1.0.5']
```
This should still be changed to development dependency. We don't want to force every selenium-webdriver user to install curb.

aguspe replied:
Oh thank you, I thought I updated this to development_dependency, let me fix it

p0deje commented on rb/spec/unit/selenium/webdriver/remote/http/curb_spec.rb:
```diff
@@ -0,0 +1,52 @@
+# under the License.
+
+require File.expand_path('../../spec_helper', __dir__)
+load 'selenium/webdriver/remote/http/curb.rb'
```
Why not `require`?

aguspe replied:
not particular reason, let me go with require

aguspe commented:
Thank you for the quick review @p0deje, ready for re-review

p0deje commented:
Looks like Curb cannot be installed on JRuby, can you move it to Gemfile similar to `debug` and make the Curb specs guarded on `platform: :java`?

aguspe commented:
> Looks like Curb cannot be installed on JRuby, can you move it to Gemfile similar to `debug` and make the Curb specs guarded on `platform: :java`?

I moved curb to the gemfile, I added an unless to the curb rspec file, because I saw we do the same in another test that cannot run in Jruby, but maybe there is a better way to do that

p0deje commented:
@aguspe Can you please take a look at the CI failures?

aguspe commented:
> @aguspe Can you please take a look at the CI failures?

I started looking at the failures today, it seems to be related to the curl library in both windows and Linux, so I will keep investigating tomorrow and ping with a possible solution 😊

p0deje commented:
@aguspe I think you need to guard loading Curb in tests on the platforms that don't support it:

```ruby
begin
  require 'curb'
rescue LoadError
  # do nothing
end
```

aguspe commented:
> ```ruby
> begin
>   require 'curb'
> rescue LoadError
>   # do nothing
> end
> ```

Thank you for the help, I also extended the guards on the tests so they will not execute it unsupported platforms

aguspe commented:
> @aguspe I think you need to guard loading Curb in tests on the platforms that don't support it:
> 
> ```ruby
> begin
>   require 'curb'
> rescue LoadError
>   # do nothing
> end
> ```

I'm trying to only target darwin as a platform for curb and removing windows and linux when the gem is being installed

luke-hill commented:
@aguspe - Installing `curb` on Windows is a non-trivial task. I was doing this at a previous job. I updated `taf2/curb` with a few rudimentary instructions on the main page README

I can take a bigger look at this later this week if I'm needed.

aguspe commented:
> @aguspe - Installing `curb` on Windows is a non-trivial task. I was doing this at a previous job. I updated `taf2/curb` with a few rudimentary instructions on the main page README
> 
> I can take a bigger look at this later this week if I'm needed.

That could be really helpful, I think for now the strategy that we will go with is to only run the tests on darwin, so we can built without breaking anything else

p0deje commented on rb/Gemfile:
```diff
@@ -5,5 +5,6 @@ Dir["#{__dir__}/*.gemspec"].each do |spec|
   gemspec name: File.basename(spec, '.gemspec')
 end
 
+gem 'curb', '~> 1.0.5', require: false, platforms: %i[darwin]
```
Why only Darwin? Curb should at least install on Linux/macOS. For Windows, I think it should be possible to install https://packages.msys2.org/package/libcurl?repo=msys&variant=x86_64 for curb to work.

aguspe replied:
@p0deje I want to see if all the pipelines failed again, because the build on ubuntu was failing due to curb, btw this line of code doesn't work neither install_if: darwin so a simple conditional will make the trick

p0deje replied:
You might need to install `libcurl` with GitHub Actions before running the build.

aguspe replied:
@p0deje I added libcurl to the bazel.yml file if the OS is linux, can we give it a try to see if this works?

aguspe replied:
Linux worked, now I'm trying on windows

p0deje replied:
For Windows, since Ruby is installed via Bazel, you might want to instruct it to install libcurl too, this is done via 

```bzl
# MODULE.bazel
ruby.toolchain(
    name = "ruby",
    version_file = "//:.ruby-version",
    msys2_packages = ["libcurl", "libyaml"]
)
```

aguspe replied:
I will give it a try to this :) thank you

aguspe replied:
I changed the package manager on github action to msys2, so hopefully this will help, it said before with bazel that it couldn't find the package

p0deje replied:
When Bazel installs Ruby on Windows, it would actually ignore system-installed MSYS2 so it would help.

I think libcurl is a wrong name, looking at https://packages.msys2.org/base/mingw-w64-curl, you probably can try install `curl` and see if it helps - Bazel will add `mingw-w64-x86_64-` prefix automatically.

aguspe commented:
The pipeline is green again, could you re-review @p0deje ? Thank you so much

p0deje commented on rb/lib/selenium/webdriver/remote/http/curb.rb:
```diff
@@ -17,7 +17,11 @@
 # under the License.
 
-require 'curb'
+begin
```
Thinking a bit more about this, we are changing behavior for end users. Previously, when the user attempted to use this file, it would fail with LoadError saying that `curb` was missing - it's a clear indication that the user needs to install `curb` manually. Now, it silently succeeds and fails later when the user attempts to use the HTTP client with a missing constant.

Let's revert this change and instead rescue LoadError inside `curb_spec.rb`.

aguspe replied:
You have a good point there, I moved the rescue into the spec, and also added a raise in case it fails for the user to know what happened and what they can do next, thank you for the review!

p0deje replied:
Sorry I might be nitpicking here, but why did you remove `require "curb"`? We want to load it so that when the user requires this HTTP client, they will see a load error that Curb is missing.

aguspe replied:
I miss understood your previous comment, let me get it back in place, you are right, thanks

p0deje commented on rb/spec/unit/selenium/webdriver/remote/http/curb_spec.rb:
```diff
@@ -0,0 +1,59 @@
+begin
+  require 'curb'
+rescue LoadError
+  raise 'Curb is not supported by this platform. Try installing Curl'
```
It's just a spec, we don't really need that.

aguspe replied:
Okey, I will roll back to the comment now

aguspe commented:
@p0deje ready for re-review

p0deje commented:
> @p0deje ready for re-review

Could you look at https://github.com/SeleniumHQ/selenium/pull/14285#discussion_r1723312179?

aguspe commented:
> rb/lib/selenium/webdriver/remote/http/curb.rb

I added the required back @p0deje thank you for spotting that

aguspe commented:
This Java test is failing, I can see that it is a new failure that was not there before, should I address it on this PR?

```
//java/test/org/openqa/selenium/bidi/input:DefaultKeyboardTest-chrome    FAILED in 2 out of 2 in 20.8s
  Stats over 2 runs: max = 20.8s, min = 20.1s, avg = 20.5s, dev = 0.3s
  /home/runner/.bazel/execroot/_main/bazel-out/k8-fastbuild/testlogs/java/test/org/openqa/selenium/bidi/input/DefaultKeyboardTest-chrome/test.log
  /home/runner/.bazel/execroot/_main/bazel-out/k8-fastbuild/testlogs/java/test/org/openqa/selenium/bidi/input/DefaultKeyboardTest-chrome/test_attempts/attempt_1.log
//rb/spec/unit/selenium/webdriver/remote/http:curb                       FAILED in 2 out of 2 in 8.3s
  Stats over 2 runs: max = 8.3s, min = 7.9s, avg = 8.1s, dev = 0.2s
  /home/runner/.bazel/execroot/_main/bazel-out/k8-fastbuild/testlogs/rb/spec/unit/selenium/webdriver/remote/http/curb/test.log
  /home/runner/.bazel/execroot/_main/bazel-out/k8-fastbuild/testlogs/rb/spec/unit/selenium/webdriver/remote/http/curb/test_attempts/attempt_1.log
  ```

p0deje commented:
@aguspe You don't need to fix the Java test, but let's fix Ruby test.

aguspe commented:
> @aguspe You don't need to fix the Java test, but let's fix Ruby test.

So the ruby test is failing due to be running on Jruby:

```
An error occurred while loading ./rb/spec/unit/selenium/webdriver/remote/http/curb_spec.rb.Failure/Error: Unable to find org/jruby/RubyKernel.java to read failed lineLoadError:  no such file to load -- curb# ./rb/lib/selenium/webdriver/remote/http/curb.rb:20:in `<main>'# ./rb/spec/unit/selenium/webdriver/remote/http/curb_spec.rb:21:in `<main>'All examples were filtered out; ignoring {}No examples found.Finished in 0.00033 seconds (files took 1.98 seconds to load)0 examples, 0 failures, 1 error occurred outside of examples
```

So I just added an unless for Jruby since in windows and linux now we have curl installed @p0deje

p0deje commented on rb/spec/unit/selenium/webdriver/remote/http/curb_spec.rb:
```diff
@@ -0,0 +1,57 @@
+# specific language governing permissions and limitations
+# under the License.
+
+require File.expand_path('../../spec_helper', __dir__)
```
Rather than having multiple guards, you can simply stop loading the file on JRuby:

```suggestion
return if Platform.jruby?
require File.expand_path('../../spec_helper', __dir__)
```

aguspe replied:
That's a better way, thank you, updated

aguspe commented:
I tried with Ruby engine locally, and the test gets ignore, is odd that it failed at the same point with Ruby Platform @p0deje

p0deje commented:
> I tried with Ruby engine locally, and the test gets ignore, is odd that it failed at the same point with Ruby Platform @p0deje

Looks like you need to guard against TruffleRuby as well. Can you try the following in spec:


```ruby
require File.expand_path('../../spec_helper', __dir__)
return if Selenium::WebDriver::Platform.jruby? || Selenium::WebDriver::Platform.truffleruby?
```

aguspe commented:
> ```ruby
> require File.expand_path('../../spec_helper', __dir__)
> return if Selenium::WebDriver::Platform.jruby? || Selenium::WebDriver::Platform.truffleruby?
> ```

Updated to support truffle ruby

aguspe commented:
Now the only test failing is the java one:

```
//java/test/org/openqa/selenium/bidi/input:DefaultKeyboardTest-chrome    FAILED in 2 out of 2 in 19.7s
  Stats over 2 runs: max = 19.7s, min = 19.2s, avg = 19.4s, dev = 0.3s
  /home/runner/.bazel/execroot/_main/bazel-out/k8-fastbuild/testlogs/java/test/org/openqa/selenium/bidi/input/DefaultKeyboardTest-chrome/test.log
  /home/runner/.bazel/execroot/_main/bazel-out/k8-fastbuild/testlogs/java/test/org/openqa/selenium/bidi/input/DefaultKeyboardTest-chrome/test_attempts/attempt_1.log

Executed 2 out of 2076 tests: 2075 tests pass and 1 fails remotely.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
```

aguspe commented:
Thanks for the help @p0deje !

p0deje commented:
@aguspe Thank you for working it out!
